plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginCommandHandler.java

```diff
@@ -1,21 +1,63 @@
 package com.google.cloud.tools.eclipse.appengine.login;
 
+import java.io.IOException;
+import java.util.Map;
+
 import org.eclipse.core.commands.AbstractHandler;
 import org.eclipse.core.commands.ExecutionEvent;
 import org.eclipse.core.commands.ExecutionException;
-import org.eclipse.ui.handlers.HandlerUtil;
 import org.eclipse.jface.dialogs.MessageDialog;
+import org.eclipse.jface.window.SameShellProvider;
 import org.eclipse.swt.widgets.Shell;
+import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.commands.ICommandService;
+import org.eclipse.ui.commands.IElementUpdater;
+import org.eclipse.ui.handlers.HandlerUtil;
+import org.eclipse.ui.internal.WorkbenchWindow;
+import org.eclipse.ui.menus.UIElement;
 
-public class GoogleLoginCommandHandler extends AbstractHandler {
+import com.google.api.client.auth.oauth2.Credential;
 
+public class GoogleLoginCommandHandler extends AbstractHandler implements IElementUpdater {
+
+  @Override
   public Object execute(ExecutionEvent event) throws ExecutionException {
     Shell shell = HandlerUtil.getActiveShell(event);
+    GoogleLoginService loginService = new GoogleLoginService();
+
+    Credential credential = loginService.getCachedActiveCredential();
+    if (credential == null) {
+      try {
+        credential = loginService.getActiveCredential(new SameShellProvider(shell));
 
-    boolean success = new GoogleLoginTemporaryTester().testLogin(shell);
+        boolean success = new GoogleLoginTemporaryTester().testLogin(credential);
+        MessageDialog.openInformation(shell,
+            "TESTING AUTH", success ? "SUCCESS" : "FAILURE (to be implemented)");
+      } catch (IOException ioe) {
+        throw new ExecutionException(ioe.getMessage());
+      }
+    } else {
+      if (MessageDialog.openConfirm(shell,
+          Messages.LOGOUT_CONFIRM_DIALOG_TITILE, Messages.LOGOUT_CONFIRM_DIALOG_MESSAGE)) {
+        loginService.clearCredential();
+      }
+    }
+
+    ICommandService commandService =
+        (ICommandService) PlatformUI.getWorkbench().getService(ICommandService.class);
+    commandService.refreshElements(
+        "com.google.cloud.tools.eclipse.appengine.login.commands.loginCommand", null); //$NON-NLS-1$
 
-    MessageDialog.openInformation(shell,
-        "TESTING AUTH", success ? "SUCCESS" : "FAILURE (to be implemented)");
     return null;
   }
+
+  @Override
+  public void updateElement(UIElement element, @SuppressWarnings("rawtypes") Map parameters) {
+    boolean loggedIn = new GoogleLoginService().getCachedActiveCredential() != null;
```

Comment:
When you get to the point of implementing an OSGi service, you can use `UIElement#getServiceLocator()`